workers/task-driver/integration/tests/pay_relayer_fee.rs

```diff
@@ -0,0 +1,189 @@
+    state.update_wallet(wallet).unwrap().await.unwrap();
+}
+
+/// Check a wallet in both the global state and by looking it up on-chain
+async fn check_result_wallet(
+    expected: Wallet,
+    blinder_seed: Scalar,
+    secret_share_seed: Scalar,
+    args: &IntegrationTestArgs,
+) -> Result<()> {
+    let state = &args.state;
+    let wallet = state.get_wallet(&expected.wallet_id)?.ok_or_else(|| eyre!("wallet not found"))?;
+
+    // We only compare the public and private shares, other metadata fields on the
+    // wallets may differ
+    assert_eq_result!(wallet.blinded_public_shares, expected.blinded_public_shares)?;
+    assert_eq_result!(wallet.private_shares, expected.private_shares)?;
+
+    // Lookup the wallet from on-chain, and verify its construction
+    lookup_wallet_and_check_result(&expected, blinder_seed, secret_share_seed, args.clone()).await
+}
+
+// --------------
+// | Test Cases |
+// --------------
+
+/// Tests paying a relayer fee when the relayer initially has no balance
+#[allow(non_snake_case)]
+async fn test_pay_relayer_fee__zero_initial_balance(test_args: IntegrationTestArgs) -> Result<()> {
+    let mut rng = thread_rng();
+
+    let blinder_seed = Scalar::random(&mut rng);
+    let share_seed = Scalar::random(&mut rng);
+    let (trader_wallet, fee_mint) =
+        setup_trader_wallet(blinder_seed, share_seed, &test_args).await?;
+
+    // Create an empty relayer wallet
+    let (relayer_wallet, relayer_blinder_seed, relayer_share_seed) =
+        new_wallet_in_darkpool(&test_args.arbitrum_client).await?;
+    set_local_relayer_wallet(relayer_wallet.clone(), &test_args).await;
+
+    // Pay the relayer fee
+    let desc = PayRelayerFeeTaskDescriptor::new(trader_wallet.wallet_id, fee_mint.clone())
+        .expect("infallible");
+    await_task(desc.into(), &test_args).await?;
+
+    // Check the trader's wallet
+    let mut expected_trader_wallet = trader_wallet.clone();
+    let amt = expected_trader_wallet.get_balance_mut(&fee_mint).unwrap().relayer_fee_balance;
+    expected_trader_wallet.get_balance_mut(&fee_mint).unwrap().relayer_fee_balance = 0;
+    expected_trader_wallet.reblind_wallet();
+
+    check_result_wallet(expected_trader_wallet, blinder_seed, share_seed, &test_args).await?;
+
+    // Check the relayer wallet
+    let mut expected_relayer_wallet = relayer_wallet.clone();
+    let bal = Balance::new_from_mint_and_amount(fee_mint, amt);
+    expected_relayer_wallet.add_balance(bal).unwrap();
+    expected_relayer_wallet.reblind_wallet();
+
+    check_result_wallet(
+        expected_relayer_wallet,
+        relayer_blinder_seed,
+        relayer_share_seed,
+        &test_args,
+    )
+    .await
+}
+integration_test_async!(test_pay_relayer_fee__zero_initial_balance);
+
+/// Tests paying a relayer fee when the relayer initially has a non-zero balance
+#[allow(non_snake_case)]
+async fn test_pay_relayer_fee__non_zero_initial_balance(
```

Comment:
seems like the only difference between the implementation of this test and the one above is the initial relayer wallet, perhaps we could absorb some of the common functionality into helpers?